### PR TITLE
feat: implement simapp and e2e tests

### LINKED ISSUE #1: Implement `SimApp` + E2E Tests

### PR DESCRIPTION
**Pull request by johnletey**
Closes #1 

## Summary by CodeRabbit

- **New Features**
  - Introduced GitHub Actions workflow for end-to-end testing with Docker image building and test matrix preparation.
  - Added local authority node setup script for blockchain networks.
  - Enhanced `SimApp` with new configurations and functionalities for various modules, including IBC modules.

- **Bug Fixes**
  - Improved handling of module parameters and upgrade scenarios through new test cases.

- **Chores**
  - Updated `.gitignore` to exclude `.authority` and `build` directories.
  - Updated `Makefile` with new targets for building, local image creation, and end-to-end testing.

- **Tests**
  - Added multiple test cases for validating client substitution, ownership transfer, parameter updates, and software upgrades in the blockchain network.

### CODE REVIEW

File: e2e/utils.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+					return cosmos.ModifyGenesis(changes)(cfg, bz)
+				},
+			},
+		},
+	}
+	if ibcEnabled {
+		specs = append(specs, &interchaintest.ChainSpec{
+			Name:          "gaia",
+			Version:       "latest",
```
Comment (boojamya) [inline e2e/utils.go]: Although unlikely, gaia could release a "latest" version that breaks our testing. Just something to consider. I usually like to pin it to a version.

File: e2e/utils.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+				Chain2:  gaia,
+				Relayer: relayer,
+				Path:    "transfer",
+			})
+	}
+	require.NoError(t, interchain.Build(ctx, execReporter, interchaintest.InterchainBuildOptions{
+		TestName:         t.Name(),
+		Client:           client,
+		NetworkID:        network,
+		SkipPathCreation: true,
```
Comment (boojamya) [inline e2e/utils.go:141]: Even if `ibcEnabled` = `true`?

Comment (boojamya) [reply e2e/utils.go:141]: Disregard, I see you are creating path manually.

File: e2e/clients.json
Comment (boojamya) [inline e2e/clients.json:1]: Is this file necessary? Should we add to `gitignore`? Or were you including these as a reference?

Same with `params.json` and `upgrade.json`

**Review by boojamya — COMMENTED** [review]
This looks great! Clean and easy to read.

Only thought is if you think it is necessary to test non happy paths such as submitting update owner, params and client transactions from non owner account.

File: e2e/utils.go (hunk 2 above)
Comment (johnletey) [reply e2e/utils.go:141]: Resolving this 🫡

File: e2e/clients.json
Comment (johnletey) [reply e2e/clients.json:1]: `clients.json` and all the other tx files are actually very important for the tests to run.

https://github.com/noble-assets/authority/blob/b20ab74be2c6fa4f4b4176c262900fcf8561c01c/e2e/clients_test.go#L16-L17

https://github.com/noble-assets/authority/blob/b20ab74be2c6fa4f4b4176c262900fcf8561c01c/e2e/clients_test.go#L61-L66

As you can see, we read in the `clients.json` file in order to execute the substitution correctly. We could of course use the transaction builder, but this is definitely the easiest way to do this IMO.

File: e2e/utils.go (hunk 1 above)
Comment (johnletey) [reply e2e/utils.go]: https://github.com/noble-assets/authority/commit/e95e3b7c69b66a5bb2aab06ede74af11650973f1

**Comment by johnletey** [comment]
Thanks for the review @boojamya!

> if you think it is necessary to test non happy paths

We should talk more about this on our internal sync later today, but it comes down to our internal reasoning around Unit vs E2E tests 🤔

File: e2e/clients.json
Comment (boojamya) [reply e2e/clients.json:1]: Yes, but doesn't it re-create the file each time the test is run, like on line 61?

**Review by boojamya — APPROVED** [review]

File: e2e/clients.json
Comment (boojamya) [reply e2e/clients.json:1]: Oh yes, ok.  I misunderstood what `WriteFile` was doing. Thanks for the explanation.

**Pull request merged**